src/cunumeric/sort/sort_cpu.inl

```diff
@@ -0,0 +1,1027 @@
+
+      // initialize indices
+      if (need_distributed_sort) {
+        auto indices_buffer  = create_buffer<int64_t>(volume);
+        indices_ptr          = indices_buffer.ptr(0);
+        local_sorted.indices = indices_buffer;
+        local_sorted.size    = volume;
+      } else {
+        AccessorWO<int64_t, DIM> output = output_array.write_accessor<int64_t, DIM>(rect);
+        assert(rect.empty() || output.accessor.is_dense_row_major(rect));
+        indices_ptr = output.ptr(rect.lo);
+      }
+      size_t offset = rect.lo[DIM - 1];
+      if (volume > 0) {
+        if (DIM == 1) {
+          thrust::sequence(exec, indices_ptr, indices_ptr + volume, offset);
+        } else {
+          thrust::transform(exec,
+                            thrust::make_counting_iterator<int64_t>(0),
+                            thrust::make_counting_iterator<int64_t>(volume),
+                            thrust::make_constant_iterator<int64_t>(segment_size_l),
+                            indices_ptr,
+                            modulusWithOffset(offset));
+        }
+      }
+    } else {
+      // initialize output
+      if (need_distributed_sort) {
+        auto input_copy      = create_buffer<VAL>(volume);
+        values_ptr           = input_copy.ptr(0);
+        local_sorted.values  = input_copy;
+        local_sorted.indices = create_buffer<int64_t>(0);
+        local_sorted.size    = volume;
+      } else {
+        AccessorWO<VAL, DIM> output = output_array.write_accessor<VAL, DIM>(rect);
+        assert(rect.empty() || output.accessor.is_dense_row_major(rect));
+        values_ptr = output.ptr(rect.lo);
+      }
+    }
+
+    if (volume > 0) {
+      // sort data (locally)
+      auto* src = input.ptr(rect.lo);
+      if (src != values_ptr) std::copy(src, src + volume, values_ptr);
+      thrust_local_sort_inplace(values_ptr, indices_ptr, volume, segment_size_l, stable, exec);
+    }
+
+    if (need_distributed_sort) {
+      if (is_index_space) {
+        assert(is_index_space || is_unbound_1d_storage);
+        std::vector<size_t> sort_ranks(num_sort_ranks);
+        size_t rank_group = local_rank / num_sort_ranks;
+        for (int r = 0; r < num_sort_ranks; ++r) sort_ranks[r] = rank_group * num_sort_ranks + r;
+
+        void* output_ptr = nullptr;
+        // in case the storage *is NOT* unbound -- we provide a target pointer
+        // in case the storage *is* unbound -- the result will be appended to output_array
+        if (volume > 0 && !is_unbound_1d_storage) {
+          if (argsort) {
+            auto output = output_array.write_accessor<int64_t, DIM>(rect);
+            assert(output.accessor.is_dense_row_major(rect));
+            output_ptr = static_cast<void*>(output.ptr(rect.lo));
+          } else {
+            auto output = output_array.write_accessor<VAL, DIM>(rect);
+            assert(output.accessor.is_dense_row_major(rect));
+            output_ptr = static_cast<void*>(output.ptr(rect.lo));
```

Comment:
lines 988-989 and 992-993 are the same, I think you can move them out of the if/else